mipengine/node/monetdb_interface/monet_db_facade.py

```diff
@@ -0,0 +1,189 @@
+from contextlib import contextmanager
+from typing import List
+
+import pymonetdb
+from eventlet.greenthread import sleep
+from eventlet.lock import Semaphore
+
+from mipengine.node import config as node_config
+from mipengine.node import node_logger as logging
+from mipengine.singleton import Singleton
+
+INTEGRITY_ERROR_RETRY_INTERVAL = 1
+BROKEN_PIPE_MAX_ATTEMPTS = 50
+BROKEN_PIPE_ERROR_RETRY = 0.2
+CREATE_OR_REPLACE_QUERY_TIMEOUT = 1
+CREATE_REMOTE_TABLE_QUERY_TIMEOUT = 1
+INSERT_INTO_QUERY_TIMEOUT = (
+    node_config.celery.worker_concurrency * node_config.celery.run_udf_time_limit
+)
+
+create_remote_table_query_lock = Semaphore()
+create_function_query_lock = Semaphore()
+insert_query_lock = Semaphore()
+
+
+class DBExecutionDTO:
+    def __init__(self, query, parameters=None, many=False):
+        self.query = query
+        self.parameters = parameters
+        self.many = many
+
+
+class _MonetDBConnectionPool(metaclass=Singleton):
+    """
+    MonetDBConnectionPool is a Singleton class.
+
+    We use pseudo-multithreading(eventlet greenlets),
+    we provide a connection pool to support concurrent query execution.
+    """
+
+    def __init__(self):
+        self._connection_pool = [self.create_connection() for _ in range(16)]
+
+    def create_connection(self):
+        return pymonetdb.connect(
+            hostname=node_config.monetdb.ip,
+            port=node_config.monetdb.port,
+            username=node_config.monetdb.username,
+            password=node_config.monetdb.password,
+            database=node_config.monetdb.database,
+        )
+
+    def get_connection(self):
+        return self._connection_pool.pop()
+
+    def release_connection(self, conn):
+        self._connection_pool.append(conn)
+
+
+@contextmanager
+def cursor(_connection):
+    """
+    We use pseudo-multithreading (eventlet greenlets) by committing before a select query we refresh the state
+    of the connection so that it sees changes from other connections.
+    https://stackoverflow.com/questions/9305669/mysql-python-connection-does-not-see-changes-to-database-made
+    """
+    _connection.commit()
+
+    cur = _connection.cursor()
+    yield cur
+    cur.close()
+
+
+def _execute_and_commit(conn, db_execution_dto):
+    with cursor(conn) as cur:
+        cur.executemany(
+            db_execution_dto.query, db_execution_dto.parameters
+        ) if db_execution_dto.many else cur.execute(
+            db_execution_dto.query, db_execution_dto.parameters
+        )
+    conn.commit()
+
+
+@contextmanager
+def _lock(query_lock, timeout):
+    query_lock.acquire(timeout=timeout)
+    yield
+    query_lock.release()
+
+
+def execute_and_fetchall(query: str, parameters=None, many=False) -> List:
+    return execute_queries_with_connection_handling(
+        func=_execute_and_fetchall,
+        query=query,
+        parameters=parameters,
+        many=many,
+    )
+
+
+def execute(query: str, parameters=None, many=False) -> List:
```

Comment:
Since this is going to be used without a prefix, rename to something visibly linked to the db?
`db_exexute` or something similar?